### PR TITLE
MH-14 [+ MH-55] create new account  backend 

### PR DESCRIPTION
**Pull request by sharrisoncaktus**
This PR creates the backend (models, view) for /accounts/signup and integrates it with the frontend (template).

## Models
- **State** – a list of all 50 states
  * name = two-letter abbreviation (RI)
  * title = common name (Rhode Island)
  * supported = whether the state is currently supported (True, None) – at present only NC and SC are supported, but this can be edited as additional states are added.

- **UserDetails** – ancillary information about the user as pertains to the MyHPOM app.
  * state (=> State) = the state the user belongs to / resides in
  * middle_name = because the User model only has first_name, last_name
  * accept_tos = whether or not the user has accepted the terms of service

## View: /accounts/signup
* **myhpom.forms.signup.SignupForm** provides the signup form data / validation.
* **POST /accounts/signup**:
  - invalid data are redisplayed with error classes and messages
  - valid signup:
    - user state is supported => redirect to myhpom:choose_network
    - user state not supported => redirect to myhpom:next_steps (skip the choose network step).

* **TODO** in a later ticket (i.e., not listed in the AC for this ticket):
  - **no user email as yet** though this is indicated 
  - **no user login as yet** though this is implied.

## Testing Notes
* the tests require the State model to be filled. So the first time you run the tests, you'll need to have it migrate the database. 
* on future runs, you can keepdb and speed things up 
* any db changes that occur are wiped within the same test case so as to keep a consistent state.

### CODE REVIEW

**Comment by dsummersl** [comment]
Can one of the admins verify this patch?

File: myhpom/models/__init__.py
Comment (dsummersl) [inline myhpom/models/__init__.py:7]: I like that this PR migrates the structure from app/models.py to app/models/a_model.py - for consistency, pull the model out of here into its own file?

Comment (dsummersl) [reply myhpom/models/__init__.py:7]: Now that I'm looking at how State/StateAdvanceDirective are related, it seems to me that this model should go away, and that the `advance_directive_template` should replace `State.supported`. I like the idea of having a follow on ticket to address this and the flow of the whole 'user signup process': make sure a user is signed in, that the duplication of these two models is addressed, the 'next steps' URL pattern no longer includes the State, and that the user lands on a dashboard page with a working logout button (so that testers can repeat tests easily). Does that sound like a reasonable follow-on ticket? If so...disregard the refactor suggestion above!

File: myhpom/templates/myhpom/accounts/signup.html
Comment (dsummersl) [inline myhpom/templates/myhpom/accounts/signup.html]: Should be uppercase as the other field description/headers are.

Comment (dsummersl) [reply myhpom/templates/myhpom/accounts/signup.html]: WRT "State of residence"

File: myhpom/templates/myhpom/accounts/signup.html
Comment (dsummersl) [inline myhpom/templates/myhpom/accounts/signup.html]: The wireframes for error validation show that the corresponding labels should also be styled in red, which isn't happening here.

Comment (dsummersl) [reply myhpom/templates/myhpom/accounts/signup.html]: The other relevant missing style here is the terms and conditions box, and title of the terms and conditions.

File: myhpom/tests/test_signup_form.py
Comment (dsummersl) [inline myhpom/tests/test_signup_form.py]: The only non-built-in part of this form is the password verification in the `clean()` method. A test for that behavior would be great here.

File: myhpom/tests/test_signup_view.py
Comment (dsummersl) [inline myhpom/tests/test_signup_view.py]: assertions on `form.errors` in the `response.context` would make sense here - we would expect no data to result in some errors!

File: myhpom/templates/myhpom/accounts/signup.html
Comment (dsummersl) [inline myhpom/templates/myhpom/accounts/signup.html]: As we're using bs4 I'm tempted to sugged we use its dropdown support and put a divider between supported states and the other states: http://getbootstrap.com/docs/4.1/components/dropdowns/

File: myhpom/tests/test_signup_view.py
Comment (dsummersl) [inline myhpom/tests/test_signup_view.py]: These `User.objects...delete()` isn't needed since you're using django's TestCase.

Comment (dsummersl) [reply myhpom/tests/test_signup_view.py]: (same for the delete() in the test case below)

File: myhpom/tests/test_signup_view.py
Comment (dsummersl) [inline myhpom/tests/test_signup_view.py]: We try to use pep8 generally as a guide for formatting, so...spaces around the `+` would be called for here. But I'd probably replace these multiple assertions with an `assertRedirects()`...like so ` self.assertRedirects(response, reverse('myhpom:choose_network'), fetch_redirect_response=False)` -- same for the test below.

File: myhpom/validators.py
Comment (dsummersl) [inline myhpom/validators.py:13]: This import is ultimately just `from django.contrib.auth.models import User` - using that directly would probably address whatever circular import required that this import get put here, and make it clearer which model User is.

File: myhpom/models/__init__.py
Comment (dsummersl) [inline myhpom/models/__init__.py:3]: Use `from django.contrib.auth.models import User` as that is what this is.

File: myhpom/models/user_details.py
Comment (dsummersl) [inline myhpom/models/user_details.py:2]: Same import comment as before...

File: myhpom/views/signup.py
Comment (dsummersl) [inline myhpom/views/signup.py:4]: Use `from django.contrib.auth.models import User` for User.

File: myhpom/models/user.py
Comment (dsummersl) [inline myhpom/models/user.py]: Document the purpose of this signal handler either by changing the method name to reflect its purpose, or with a docstring - I assume its ultimately to populate the username field, but I'm not sure.

File: myhpom/validators.py
Comment (dsummersl) [inline myhpom/validators.py]: A user might construe the message to mean they only need a password with whatever validation warnings are currently displayed and cycle between their favorites of, say, "CarBaby!" (Please enter a password with at least 1 number (0-9)
) and "BornIn04" (Please enter a password with at least 1 special character...) and be just a tad confused what they need to do.

Perhaps..."Your password needs at least "?

File: myhpom/views/signup.py
Comment (dsummersl) [inline myhpom/views/signup.py]: `is True` is more pythonic.

File: myhpom/views/signup.py
Comment (dsummersl) [inline myhpom/views/signup.py:9]: one TODO looks to be addressed already, so that comment could go (email already taken)

File: myhpom/tests/test_signup_view.py
Comment (dsummersl) [inline myhpom/tests/test_signup_view.py:1]: Since you're changing the general structure of how tests were done before this PR, migrate the remaining test that doesn't follow that pattern...test_views.py to test_next_steps_view.py?

File: myhpom/models/user.py
Comment (dsummersl) [reply myhpom/models/user.py]: For future maintainers it would be very helpful to have a test_user_model.py that captures the intended use of this signal handler.

File: myhpom/migrations/0003_user_details.py
Comment (dsummersl) [inline myhpom/migrations/0003_user_details.py:19]: You expressed the need to not reuse UserProfile and I agree...the less we're entangled with the hydroshare codebase the better!

**Review by dsummersl — CHANGES_REQUESTED** [review]
Sorry for the big # of comments (its a big PR!) - I'm going to go ahead and make that follow-on ticket to address the overlapping intent of State/StateAdvanceDirective unless I hear an objection (ie, you *really* want to address it in the PR instead!)

**Comment by sharrisoncaktus** [comment]
Thanks for the *good* number of comments – I would rather that than have
things not addressed, and I'll be in novitiate mode for awhile. :-)

Let's finish up this PR directly – I'll make the adjustments you've
suggested to it except for integration with the next_steps.

File: myhpom/models/__init__.py
Comment (sharrisoncaktus) [reply myhpom/models/__init__.py:7]: (defer to MH-58) State.supported is currently being used for sorting purposes, so if that column is replaced the sorting logic will also need to be addressed.

File: myhpom/tests/test_signup_view.py
Comment (sharrisoncaktus) [reply myhpom/tests/test_signup_view.py]: Doing it this way enables us to keepdb / with migrations, right? We could move it to tearDown, but it seems necessary to include it if we're going to test via `REUSE_DB=1 python manage.py test --nomigrations --keepdb`

File: myhpom/templates/myhpom/accounts/signup.html
Comment (sharrisoncaktus) [reply myhpom/templates/myhpom/accounts/signup.html]: I would like to defer this change.

File: myhpom/tests/test_signup_view.py
Comment (sharrisoncaktus) [reply myhpom/tests/test_signup_view.py]: makes sense. BTW, I usually format code with `black -l 100`

File: myhpom/templates/myhpom/accounts/signup.html
Comment (dsummersl) [reply myhpom/templates/myhpom/accounts/signup.html]: np - I may well be over-anticipating what they want.

File: myhpom/tests/test_signup_view.py
Comment (sharrisoncaktus) [reply myhpom/tests/test_signup_view.py]: If we're going to `REUSE_DB=1 python manage.py test --nomigrations --keepdb`, I would think we would need to tear down whatever we add to the database.

Comment (dsummersl) [reply myhpom/tests/test_signup_view.py]: seemed to work fine when I tested it manually, which I think makes sense - afaik TestCase will wrap the tests in transactions, and roll back any modification to the database on teardown so that you don't have to...

File: myhpom/tests/test_signup_view.py
Comment (dsummersl) [reply myhpom/tests/test_signup_view.py]: Ah great; I've encountered black a few times, and have yet to use it.

File: myhpom/validators.py
Comment (sharrisoncaktus) [reply myhpom/validators.py:13]: I could do that, but the reasons for importing myhpom.models.user.User everywhere are (1) we ensure that the pre_save is picked up, and (2) if we create our own subclass later, we don't have to edit every reference to the User model. 

The pre_save is using validators, which is why we end up with a circular commit. Maybe it's not necessary to do the pre_save validation, but I like the added safety. In which case, one of the imports has to be at runtime rather than import time.

File: myhpom/models/user.py
Comment (sharrisoncaktus) [reply myhpom/models/user.py]: Good idea – I've done both

File: myhpom/validators.py
Comment (sharrisoncaktus) [reply myhpom/validators.py]: I've simplified the password_validator to give one message no matter what error occurred.

File: myhpom/tests/test_signup_view.py
Comment (sharrisoncaktus) [reply myhpom/tests/test_signup_view.py:1]: good change.

File: myhpom/migrations/0003_user_details.py
Comment (sharrisoncaktus) [reply myhpom/migrations/0003_user_details.py:19]: +1

File: myhpom/tests/test_signup_view.py
Comment (kmtracey) [reply myhpom/tests/test_signup_view.py]: Right, --keepdb avoids the overhead of the initial creation/migration of the DB, by keeping the test database around for reuse after the full test run completes. Between every test case the action of the previous test case are undone (usually by rolling back the transaction the test case was run in).  --keepdb doesn't affect what happens between test cases in a single test run.

File: myhpom/forms/signup.py
Comment (dsummersl) [inline myhpom/forms/signup.py]: Do you need to store `cleaned_data` in the form class? This could be as simple as:

```
    def clean(self):
        cleaned_data = super(SignupForm, self).clean()
        password = cleaned_data.get("password")
        password_confirm = cleaned_data.get("password_confirm")
        if password != password_confirm:
            self.add_error('password_confirm', "Please enter exactly the same password again")

        return cleaned_data
```

In manually testing this if I enter no password for the the password_confirm field I end up getting two errors on the field: one from the builtin field’s error_messages, and one from this clean() method. To fix that you probably need to inspect `self.errors` and refrain from adding an error for password_confirm if there already is one.

File: myhpom/static/myhpom/page/_signup.scss
Comment (dsummersl) [inline myhpom/static/myhpom/page/_signup.scss]: Create a variable for this color, since its repeated a few times. I imagine as we get more from Shane we'll want to make our own theme for bootstrap, but for the time being....defining the magic color codes used more than once seems sensible to me...

File: myhpom/templates/myhpom/accounts/signup.html
Comment (dsummersl) [inline myhpom/templates/myhpom/accounts/signup.html]: When `form.errors.accept_tos` is true, there are now two boxes: one from `form-group-tos__content` (gray) and one from this new class (red). I'd suggest moving this if-condition up one level in the DOM and renaming it `form-group-tos__content--invalid`, and switch between these two box classes.

**Review by dsummersl — CHANGES_REQUESTED** [review]
Still some stragglers here, and I've found a couple issues (see comments). I also had some trouble getting the styles to 'take' right away. But after I ran:

`./hsctl managepy compilescss`

I saw the label styling changes fine (+1 to that part).

File: myhpom/tests/test_signup_view.py
Comment (sharrisoncaktus) [reply myhpom/tests/test_signup_view.py]: Got it, thanks.

File: myhpom/forms/signup.py
Comment (sharrisoncaktus) [reply myhpom/forms/signup.py]: +1

File: myhpom/static/myhpom/page/_signup.scss
Comment (sharrisoncaktus) [reply myhpom/static/myhpom/page/_signup.scss]: good idea.

**Review by dsummersl — APPROVED** [review]
We're there! 🎀

**Pull request merged**
